### PR TITLE
[libc++] Use GitHub-provided runners for the windows CI

### PR DESCRIPTION
**Pull request by philnik777**
Co-authored-by: Martin Storsjö <martin@martin.st>

### CODE REVIEW

**Comment by philnik777** [comment]
@mstorsjo Could you take a look here? I've tried a bunch of stuff, but couldn't really get things to run properly.

**Comment by mstorsjo** [comment]
> > @mstorsjo Could you take a look here? I've tried a bunch of stuff, but couldn't really get things to run properly.
> 
> Sure, I can try to set this up. I have tested running libcxx tests on Windows GitHub Actions runners before, and got it working before (for a subset of configs, not using the libcxx `run-buildbot` script) - I'll see what I can do.

I got it working for all of our Windows CI configs. See the two topmost commits at https://github.com/mstorsjo/llvm-project/commits/libcxx-gh-win-runners. The one setting up the actions stuff is quite straightforward. The other one fixes a seriously annoying issue in how CMake picks python, please read the commit message thoroughly.

Also, running the libcxx tests on Github runners used to be quite annoying as they used to be quite limited (2 cores, 8 GB memory only), so earlier, I've had to set `-DLIBCXX_EXTRA_SITE_DEFINES="TEST_IS_EXECUTED_IN_A_SLOW_ENVIRONMENT"` and `-DLIBCXX_TEST_PARAMS="large_tests=False"` to make it work reliably - see https://github.com/mstorsjo/llvm-mingw/blob/cfe2ac2783dfd86a7e9ae81aba2f5db3f1f64133/.github/workflows/test-libcxx.yml#L97-L98. But with the new runners it seems to run well without that (although I haven't run all that many rounds yet, so perhaps it fails sporadically still, who knows?).

File: .github/workflows/libcxx-build-and-test.yaml [hunk 1]
```diff
@@ -199,3 +199,44 @@ jobs:
             **/CMakeError.log
             **/CMakeOutput.log
             **/crash_diagnostics/*
+  windows:
```
Comment (ldionne) [inline .github/workflows/libcxx-build-and-test.yaml:202]: When we're ready to merge this, we should add `needs: [ stage1, stage2, stage3 ]` to only trigger the Windows CI if the other stages are happy.

**Review by ldionne — APPROVED** [review]
TBH I think this LGTM if the CI passes.

I do expect that we'll hit resource limitations pretty quickly, but I guess we can cross that bridge when we get to it. This might entail setting up self-hosted runners for Windows.

**Comment by tstellar** [comment]
Can you limit the total number of jobs to 40?  We need to have some spare capacity for running some of the other jobs, like the PR labeler, code_formatter, etc.  We may be able to bump it higher in the future, but I think 40 is a good start.

**Comment by tstellar** [comment]
> Can you limit the total number of jobs to 40? We need to have some spare capacity for running some of the other jobs, like the PR labeler, code_formatter, etc. We may be able to bump it higher in the future, but I think 40 is a good start.

I thought this was possible, but I haven't found a way to actually limit the number of jobs.  I f you know how to do implement this then that is great, but otherwise do you have an estimate of the number of concurrent machines you will need to handle the peak load?

**Comment by philnik777** [comment]
> > Can you limit the total number of jobs to 40? We need to have some spare capacity for running some of the other jobs, like the PR labeler, code_formatter, etc. We may be able to bump it higher in the future, but I think 40 is a good start.
> 
> I thought this was possible, but I haven't found a way to actually limit the number of jobs. I f you know how to implement this then that is great, but otherwise do you have an estimate of the number of concurrent machines you will need to handle the peak load?

Hm, yeah. I thought it could be done with `max-parallel:`, but that only limits the parallelism per PR. I'd expect a maximum of ~10 runs/hour for libc++, which would be 80 runners. That would only be the case once in a blue moon though. A lot of the work on libc++ is done off-peak and we generally don't throw out a bunch of PRs at once. If this actually becomes a problem we could also consider moving only some of the jobs to GitHub or limiting the number of parallel jobs to 4/PR.

**Comment by tstellar** [comment]
@philnik777 OK, let's just keep it as-is and see how it goes.

File: runtimes/CMakeLists.txt [hunk 2]
```diff
@@ -155,6 +155,9 @@ set(LLVM_COMPILER_CHECKED ON)
 include(AddLLVM)
 include(HandleLLVMOptions)
 
+# Loot at the PATH first to avoid a version mismatch between the command-line
+# python and the CMake-found version
+set(Python3_FIND_REGISTRY LAST)
```
Comment (mstorsjo) [inline runtimes/CMakeLists.txt:160]: FYI - ideally we wouldn't need to have things like this in the main CMakeLists.txt.

Ideally we'd set it in the CI environment, as we happen to need it there due to how the Github runners are set up. It is possible to pass this on the CMake command line, I think, like `-DPython3_FIND_REGISTRY=LAST`, but the `run-buildbot` script doesn't support passing extra command line parameters to cmake.

On the other hand - this doesn't really hurt either; the cmake default behaviour is really really surprising here, and this makes it less surprising. It makes things a little inconsistent to keep setting this in the `runtimes/CMakeLists.txt` only - I wonder if we could set it e.g. `cmake/Modules/CMakePolicy.cmake` as a sort of project-wide default policy?

Anyway, just venting my thoughts on the matter - no urgent action needed.

**Pull request merged**